[김미소]Week4

nightowlzz opened the pull request:
## 요구사항

### 기본

- [x] [기본]이메일 input에서 focus out 할 때, 값이 없을 경우 input에 빨강색 테두리와 아래에 “이메일을 입력해주세요.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]이메일 input에서 focus out 할 때, 이메일 형식에 맞지 않는 값이 있는 경우 input에 빨강색 테두리와 아래에 “올바른 이메일 주소가 아닙니다.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]이메일: test@codeit.com, 비밀번호: codeit101 으로 로그인 시도할 경우, “/folder” 페이지로 이동하나요?
- [x] [기본]비밀번호 input에서 focus out 할 때, 값이 없을 경우 아래에 “비밀번호를 입력해주세요.” 에러 메세지가 보이나요?
- [x] [기본]이외의 로그인 시도의 경우 이메일, 비밀번호 input에 빨강색 테두리와 각각의 input아래에 “이메일을 확인해주세요.”, “비밀번호를 확인해주세요.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]이메일, 비밀번호 input에 에러 관련 디자인을 Components 영역의 에러 케이스로 적용했나요?
- [x] [기본]로그인 버튼 클릭 또는 Enter키 입력으로 로그인 되나요?

### 심화

- [x] [심화]눈 모양 아이콘 클릭시 비밀번호의 문자열이 보이기도 하고, 가려지기도 하나요?
- [x] [심화]비밀번호의 문자열이 가려질 때는 눈 모양 아이콘에는 사선이 그어져있고, 비밀번호의 문자열이 보일 때는 사선이 없는 눈 모양 아이콘이 보이나요?

### 추가작업
- [x] [추가] input 모두 입력하고, 유효성 검사 맞으면 버튼 활성화 되도록 작업
- [x] [추가] 회원가입 input 작성하고, 회원가입 클릭 시 로그인 페이지로 이동

## 주요 변경사항

- **회원가입, 로그인 페이지** validation 작업(에러 문구, 라인)
- 눈모양 버튼 클릭시. 비밀번호 보이고 안보이고 작업

## 스크린샷

[피그마주소 week4 디자인>](https://www.figma.com/file/dH3gN96jnAtMjxjhV80HM2/Weekly-Mission%5B5%EA%B8%B0%5D?type=design&node-id=439-2193&mode=design&t=slaznPjtEVmjv29z-0)


## 멘토에게

- 잘 부탁 드립니다~

nightowlzz committed:
Feat: 회원가입 validation 작업

nightowlzz committed:
Refactor: 패스워드 validation기능을 validate.js로 이동 및 password.js수정

nightowlzz committed:
Feat: 로그인 특정아이디 입력시 이동 및 에러 추가

nightowlzz committed:
Fix: 파일이름 변경 및 로그인 validation 기능 추가

nightowlzz committed:
Fix: 'login'문자 'signin'으로 전체 변경

nightowlzz committed:
Fix: validation.js=>signup.js 이름변경 및 validation 회원가입만 적용

nightowlzz committed:
Fix: 로그인 validation 수정작업

nightowlzz committed:
Fix: input 빈공간 제거

Siihyun commented on js/password_toggle.js:
```diff
@@ -0,0 +1,19 @@
+const btnTypeChange = document.querySelectorAll(".form__box.pw button"); // join pass2 button
```
1. btnTypeChange라는 변수명은 의미를 적절하게 담지 못하는 것 같아요. passwordIcons, passwordIconButtons 등으로 수정해보는건 어떤가요?
2. .form__box.pw button 보다는, 직접 가지고 있는 class인 btn__pw을 활용해서 가져올 순 없나요?

Siihyun commented on js/password_toggle.js:
```diff
@@ -0,0 +1,19 @@
+
+function changeInputType (e) {
+  const btn = e.currentTarget;
+  const parant = e.target.closest('.form__box')
+  let inputType = parant.querySelector('input');
```
parent가 굳이 필요 없을 것 같은데, e.target.closest('#form__pw')로 가져올 순 없을까요?

nightowlzz replied:
e.target.closest('.form__box')로 사용하겠습니다!

Siihyun commented on js/signin.js:
```diff
@@ -0,0 +1,92 @@
+const signinForm = document.querySelector('.signin__wrap form');
+const signinFomeInput = document.querySelectorAll('.signin__wrap input');
+const signinEmail = document.querySelector('.signin__wrap #signin__email');
+const signinPassword = document.querySelector('.signin__wrap #signin__pw__input');
+const submitButton = document.querySelector('.signin__wrap .btn_signin');
+let validateObject = { // 활성화
+  email:false,
+  password:false,
+}; 
+
```
현재 모든 element가 전역 변수(파일 내 전체에서 접근 가능한 변수)로 관리되고 있는데, 이는 좋은 설계가 아니에요.

파일이 길어지고 함수가 많아지면, 여기저기서 signIn form이나, input, button등에 전부 접근하게 되고 값의 변경이 어디서 일어났는지 정확히 파악하기 어려워지기 때문이에요.

validateForm 안에서만 해당 변수를 가져와서 제어할 수 있도록 하고, validateForm 내부에서 호출되는 함수에서는 값을 받아서 수정하는 방향으로 바꿔보는게 어떨까요?

nightowlzz replied:
어떤식으로 작업을 해야 하는지 전혀 감이 오지 않습니다 ㅠㅠ 
전역으로 사용하지 않으면 어떤 방식으로 작업을 해야 하나요??

nightowlzz replied:
어떻게 해야하는지 모르겟어요 ㅠㅠ

Siihyun commented on js/signin.js:
```diff
@@ -0,0 +1,92 @@
+  }
+}
+
+function validatePassword () { // 비밀번호
+  if(signinPassword.value.trim() !== '') {
+    validateObject.password = true;
+    errorInputRemove(signinPassword);
+  } else {
+    validateObject.password = false;
+    errorInputAdd(signinPassword, '비밀번호를 입력해 주세요.');
+  }
+}
```
하나의 함수가 두개 이상의 일을 하고있어요.
1. password 검증
2. errorMsg 추가

이 함수는 단순 검증의 역할만 수행하고, 아래에서 validateForm 함수에서 아래와 같이 동작하도록 바꿔보면 어떨까요?

`if(validatePassword()) errorInputAdd();`

Siihyun commented on js/signin.js:
```diff
@@ -0,0 +1,92 @@
+  parent.querySelector('.error__text').innerText = massage;
+}
+
+function errorInputRemove (target) { // error massage remove
```
removeInputMessage와 같은 함수명으로 바꿔보는건 어떨까요?

Siihyun reviewed (commented):
고생하셨습니다!
커멘트 드린 부분 수정해보면 좋을 것 같아요 :)

nightowlzz committed:
Fix: 변수명 수정

nightowlzz committed:
Refactor: 여러 기능을 수행하던 함수들을 단일 기능으로 함수 분리

'emailErrorMessage','passwordErrorMessage'를 추가하여,
'validateEmail','validatePassword'에서 에러메세지 기능을 따로
분리하였습니다.

nightowlzz committed:
Refactor: join.js처럼 여러기능함수를 단일기능함수로 수정
